Implemented API test for capsule sync multiple puppet module versions(BZ1365952)

abalakh opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1365952
```python
py.test -v tests/foreman/api/test_contentmanagement.py -k test_positive_sync_puppet_module_with_versions
============================= test session starts ==============================
platform darwin -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0 -- /Users/andrii/workspace/env/bin/python
cachedir: .cache
rootdir: /Users/andrii/workspace/robottelo, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.2, forked-0.2, cov-2.5.1
collected 10 items
2017-11-03 16:38:54 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/api/test_contentmanagement.py::CapsuleContentManagementTestCase::test_positive_sync_puppet_module_with_versions PASSED

============================== 9 tests deselected ==============================
================== 1 passed, 9 deselected in 2188.99 seconds ===================
```

abalakh commented:
@oshtaier please point what do you mean by 'manual sync procedure' - `capsule.content_get_sync()` is just a GET request to fetch sync status, it's not manual trigger. Or which command do you consider as manual trigger?

oshtaier commented:
Ohh, I thought that `puppet_repository.sync()` should be automatic too. Sorry, skip my comment

ldjebran commented on tests/foreman/api/test_contentmanagement.py:
```diff
@@ -932,6 +933,122 @@ def test_positive_update_with_immediate_sync(self):
+        content_view = entities.ContentView(organization=org).create()
+        prod = entities.Product(organization=org).create()
+        puppet_repository = entities.Repository(
+            content_type='puppet',
```
Please use REPO_TYPE['puppet']

ldjebran commented on tests/foreman/api/test_contentmanagement.py:
```diff
@@ -932,6 +933,122 @@ def test_positive_update_with_immediate_sync(self):
+        else:
+            for task in sync_status['active_sync_tasks']:
+                tasks.append(entities.ForemanTask(id=task['id']))
+                tasks[-1].poll()
```
Think we do not need tasks list (perhaps affected by an early idea)

abalakh commented:
@ldjebran addressed comments. Regarding test_positive_add_puppet_module - that was small hidden bug which wasn't affecting us yet - the puppet repo has to be synced first to be able to list modules. But as tests are reusing the same repo, one of tests which is scheduled before is doing the sync, so the issue is not occurring on jenkins. However, if you run `test_positive_add_puppet_module` locally on fresh sat box, it will fail.
Test results:
```python
py.test -v tests/foreman/cli/test_contentview.py -k test_positive_add_puppet_module
============================= test session starts ==============================
platform darwin -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0 -- /Users/andrii/workspace/env/bin/python
cachedir: .cache
rootdir: /Users/andrii/workspace/robottelo, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.2, forked-0.2, cov-2.5.1
collected 96 items
2017-11-07 20:03:47 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/cli/test_contentview.py::ContentViewTestCase::test_positive_add_puppet_module <- robottelo/decorators/__init__.py PASSED
tests/foreman/cli/test_contentview.py::ContentViewTestCase::test_positive_add_puppet_module_older_version <- robottelo/decorators/__init__.py PASSED

============================= 94 tests deselected ==============================
================== 2 passed, 94 deselected in 149.61 seconds ===================
```

abalakh commented:
```python
py.test -v tests/foreman/api/test_contentmanagement.py -k test_positive_sync_puppet_module_with_versions
============================= test session starts ==============================
platform darwin -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0 -- /Users/andrii/workspace/env/bin/python
cachedir: .cache
rootdir: /Users/andrii/workspace/robottelo, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.2, forked-0.2, cov-2.5.1
collected 10 items
2017-11-08 18:09:01 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/api/test_contentmanagement.py::CapsuleContentManagementTestCase::test_positive_sync_puppet_module_with_versions PASSED

============================== 9 tests deselected ==============================
=================== 1 passed, 9 deselected in 79.03 seconds ====================
```

oshtaier commented on tests/foreman/api/test_contentmanagement.py:
```diff
@@ -932,6 +934,120 @@ def test_positive_update_with_immediate_sync(self):
+        # failed tasks were logged for capsule
+        sync_status = capsule.content_get_sync()
+        sync_errors_after = sync_status['last_failed_sync_tasks']
+        self.assertEqual(len(sync_errors_before), len(sync_errors_after))
```
Could we check that actually last version of module present on capsule? Or what that sync procedure does?

abalakh replied:
Yep, we can. BZ was about inability to perform sync successfully, and also we have multiple tests which are checking whether there's latest version of package available after sync, that why i skipped such check for this specific test. Want me to add it?

oshtaier replied:
Then honestly, I don't understand why do we need all this code at all:
```
+        # Wait till capsule sync finishes
 +        sync_status = capsule.content_get_sync()
 +        if sync_status['active_sync_tasks']:
 +            for task in sync_status['active_sync_tasks']:
 +                entities.ForemanTask(id=task['id']).poll()
 +        else:
 +            self.assertNotEqual(
 +                sync_status['last_sync_time'], last_sync_time)
 +        # Polling the task will raise an assertion in case of error, but in
 +        # case we couldn't catch tasks running in time, let's ensure no new
 +        # failed tasks were logged for capsule
 +        sync_status = capsule.content_get_sync()
 +        sync_errors_after = sync_status['last_failed_sync_tasks']
+         self.assertEqual(len(sync_errors_before), len(sync_errors_after))
```
just check that action was really did what is has to do

abalakh commented:
@oshtaier addressed in a separate commit, please revisit
```python
py.test -v tests/foreman/api/test_contentmanagement.py -k test_positive_sync_puppet_module_with_versions
============================= test session starts ==============================
platform darwin -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0 -- /Users/andrii/workspace/env/bin/python
cachedir: .cache
rootdir: /Users/andrii/workspace/robottelo, inifile:
Verifying puppet module is present on capsule
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.2, forked-0.2, cov-2.5.1
collected 10 items
2017-11-21 19:09:18 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/api/test_contentmanagement.py::CapsuleContentManagementTestCase::test_positive_sync_puppet_module_with_versions PASSED

============================== 9 tests deselected ==============================
=================== 1 passed, 9 deselected in 93.81 seconds ====================
```
